compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -124,12 +124,14 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             DynStar => {
                 if let ty::Dynamic(data, _, ty::DynStar) = cast_ty.kind() {
                     // Initial cast from sized to dyn trait
+                    // Copy the data into the first field of the destination.
+                    let data_dest = self.place_field(dest, 0)?;
+                    self.copy_op(src, &data_dest, true)?;
+                    // Copy the vtable into the second field of the destination.
                     let vtable = self.get_vtable_ptr(src.layout.ty, data.principal())?;
                     let vtable = Scalar::from_maybe_pointer(vtable, self);
-                    let data = self.read_immediate(src)?.to_scalar();
-                    let _assert_pointer_sized = data.to_pointer(self)?;
-                    let val = Immediate::ScalarPair(data, vtable);
-                    self.write_immediate(val, dest)?;
+                    let vtable_dest = self.place_field(dest, 1)?;
+                    self.write_scalar(vtable, &vtable_dest)?;
```

Comment:
What would be a test that covers this?

Reply:
The test I changed from check-pass to build-pass ICEs at the `read_immediate` I removed, so that exercises these changes I guess.

Reply:
Uh, that does not sound very confident.^^ Also that test doesn't use `const` nor Miri so it cannot be a proper test for this.

Reply:
I'll commit a Miri version of the test below then.

Reply:
You can add that test in this PR.

Reply:
Yeah I agree. Forgot to comment that I tried and failed to test a ScalarPair.